### ISSUE TITLE
Is there a bug in `FeatureStore` + `NegativeSampling`?

### ISSUE THREAD

**Issue by denadai2** [pyg-team/pytorch_geometric#8886]
### 🐛 Describe the bug

I am using FeatureStore in a distributed large scale setting. I find that features in negatives are **sometimes** different from expected. Is there a bug?

```
import torch
from torch_geometric.sampler import NegativeSampling

from torch_geometric.data import HeteroData
from torch_geometric.loader import LinkNeighborLoader
from torch_geometric.testing import (
    MyFeatureStore,
    MyGraphStore,
    get_random_edge_index,
)


def test_custom_hetero_link_neighbor_loader():
    data = HeteroData()
    feature_store = MyFeatureStore()
    graph_store = MyGraphStore()

    # Set up node features:
    x = torch.arange(10)
    data['paper'].x = x
    feature_store.put_tensor(x, group_name='paper', attr_name='x', index=None)

    x = torch.arange(10, 300)
    data['author'].x = x
    feature_store.put_tensor(x, group_name='author', attr_name='x', index=None)

    # Set up edge indices (GraphStore does not support `edge_attr` at the
    # moment):
    edge_index = get_random_edge_index(10, 10, 50)
    data['paper', 'to', 'paper'].edge_index = edge_index
    graph_store.put_edge_index(edge_index=(edge_index[0], edge_index[1]),
                               edge_type=('paper', 'to', 'paper'),
                               layout='coo', size=(10, 10))

    edge_index = get_random_edge_index(10, 20, 100)
    data['paper', 'to', 'author'].edge_index = edge_index
    graph_store.put_edge_index(edge_index=(edge_index[0], edge_index[1]),
                               edge_type=('paper', 'to', 'author'),
                               layout='coo', size=(10, 20))

    edge_index = get_random_edge_index(20, 10, 100)
    data['author', 'to', 'paper'].edge_index = edge_index
    graph_store.put_edge_index(edge_index=(edge_index[0], edge_index[1]),
                               edge_type=('author', 'to', 'paper'),
                               layout='coo', size=(20, 10))

    loader1 = LinkNeighborLoader(
        data,
        num_neighbors=[-1] * 2,
        edge_label_index=('paper', 'to', 'author'),
        batch_size=20,
        neg_sampling= NegativeSampling(
            mode="triplet", amount=1)
    )

    loader2 = LinkNeighborLoader(
        (feature_store, graph_store),
        num_neighbors=[-1] * 2,
        edge_label_index=('paper', 'to', 'author'),
        batch_size=20,
        neg_sampling= NegativeSampling(
        mode="triplet", amount=1
    )
    )

    assert str(loader1) == str(loader2)

    for (batch1, batch2) in zip(loader1, loader2):
        print("negative node ids batch1", batch1['author'].dst_neg_index)
        print("negative node ids batch2", batch2['author'].dst_neg_index)

        common_indices = list(set(batch1['author'].dst_neg_index.tolist()).intersection(set(batch2['author'].dst_neg_index.tolist())))
        # get just one common node id
        common_indices = common_indices[0]
        print("common node id in negatives", common_indices)

        mask_common_index1 = batch1["author"].dst_neg_index == common_indices
        mask_common_index2 = batch2["author"].dst_neg_index == common_indices

        # node ids
        # and let's limit to one element only (there might be more than one match)
        node_id1 = batch1["author"].dst_neg_index[mask_common_index1][0]
        node_id2 = batch2["author"].dst_neg_index[mask_common_index2][0]

        # PASSES!
        assert node_id2 == node_id1

        print(batch1["author"].x)
        print(batch1["author"].x[node_id1])
        print(batch2["author"].x[node_id2])

        # NOPE! (sometimes, so run it multiple times)
        assert batch1["author"].x[node_id1] == batch2["author"].x[node_id2]

```
output

```
negative node ids batch1 tensor([28,  9, 14, 27, 21, 18, 29, 30, 25, 20, 16,  7, 22, 23, 24,  5, 19, 26,
        15, 17])
negative node ids batch2 tensor([11, 12,  5,  2,  9, 14,  0,  7,  0,  4,  3,  5,  3,  4, 11,  7, 15,  3,
        16, 15])
common node id in negatives 5
tensor([ 10,  11,  13,  14,  16,  17,  18,  19,  21,  22,  25,  26,  27,  29,
         45,  56,  65,  73,  77,  81,  93,  98, 108, 127, 138, 176, 186, 193,
        209, 279, 281,  23,  15,  28,  12,  20,  24])
tensor(17)
tensor(16)
====================================================================== short test summary info =======================================================================
FAILED tests/ray_utils/test_ll.py::test_custom_hetero_link_neighbor_loader - assert tensor(17) == tensor(16)
```

### Versions

PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.3 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.9.16 (main, Mar  8 2023, 04:29:24)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.3-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] mypy==1.7.0
[pip3] mypy-extensions==1.0.0
[pip3] mypy-protobuf==1.23
[pip3] numpy==1.23.5
[pip3] torch==2.0.0
[pip3] torch-cluster==1.6.1
[pip3] torch-geometric==2.3.1
[pip3] torch-scatter==2.1.1
[pip3] torch-sparse==0.6.17
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.2
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] torch-cluster             1.6.1                    pypi_0    pypi
[conda] torch-geometric           2.3.1                    pypi_0    pypi
[conda] torch-scatter             2.1.1                    pypi_0    pypi
[conda] torch-sparse              0.6.17                   pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi

**Comment by rusty1s** [comment]
The negative sampling algorithm performs random sampling, so it is expected that the set of negatives is different for two different loaders.

**Comment by denadai2** [comment]
I might be wrong here but the code here above does:

1. gets the set of negatives from the two loaders
2. Gets the common negative indices (bc of the random behaviour you described)
3. Asserts the common indices are really common ahah sanity check
4. Asserts the features of these indices are the same. Nope!

**Comment by rusty1s** [comment]
Got it, I misread, sorry :)

Note that the indices returned by the loader do not refer to global indices, but dictate local indices specific to the sampled subgraph. That is, a shared index in `dst_neg_index` does not necessarily mean that it is referring to the same global node in two separate subgraphs. Here is a correct example:
```python
        global1 = batch1['author'].n_id[batch1['author'].dst_neg_index]
        global2 = batch2['author'].n_id[batch2['author'].dst_neg_index]

        mask = torch.isin(global1, global2)
        common_index = global1[mask][0]

        mask_common_index1 = global1 == common_index
        mask_common_index2 = global2 == common_index

        node_id1 = batch1["author"].dst_neg_index[mask_common_index1][0]
        node_id2 = batch2["author"].dst_neg_index[mask_common_index2][0]

        assert (batch1["author"].x[node_id1] == batch2["author"].x[node_id2])

**Comment by denadai2** [comment]
damn :D sorry and thx!
